src/webui/api/torrentscontroller.cpp

```diff
@@ -560,6 +579,84 @@ void TorrentsController::webseedsAction()
     setResult(webSeedList);
 }
 
+void TorrentsController::addWebseedsAction()
+{
+    requireParams({u"hash"_s, u"urls"_s});
+    const QStringList paramUrls = params()[u"urls"_s].split(u'|', Qt::SkipEmptyParts);
+
+    const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
+    BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QList<QUrl> urls;
+    urls.reserve(paramUrls.size());
+    for (const QString &urlStr : paramUrls)
+    {
+        const auto result = validateWebseedUrl(QUrl::fromPercentEncoding(urlStr.toUtf8()));
```

Comment:
Why not move "percent decoding" step into `validateWebseedUrl()`?

Reply:
Will do